### PR TITLE
[RFC] Release/2.0.0

### PR DESCRIPTION
**Pull request by discordier**
This PR is the work in progress for the new version `2.0` which will provide compatibility with Contao 3 & 4.

### CODE REVIEW

**Comment by dmolineus** [comment]
Looks good so far. What do you think about simplifying the definition of fallback services for Contao 3 by adding a second param to the share `method`:

```php
<?php

$container['example'] = $container->share(
    function () {
        return new Example();
    },
    'vendor.example'
);

```

Inside the share method it's checked if the symfony container is available. If the container is available and second param is passed, the service is delegated. Overthise the first param is used as fallback (Contao 3). 

Well, the definition of the closures are parsed in Contao 4 as well. I wouldn't mind for that overhead.

**Comment by discordier** [comment]
Problem is, the `share` method is stateless and therefore we would then always be setting a service, the alias will get ignored or, worse the exception that the service has been overridden will get triggered.

So I see no way how your idea could work, if you see any, please explain as I'm always open for simplifying stuff.

**Comment by baumannsven** [comment]
I have update this for contao 4.

For travis i must install contao manager plugin or excluded the path?

@discordier Can you show this?

File: composer.json [hunk 1]
```diff
@@ -39,8 +39,8 @@
     "contao/manager-bundle": "~4.4"
   },
   "replace": {
-    "contao-legacy/dependency-container": "self.version",
-    "bit3/contao-dependency-container": "self.version"
+    "contao-legacy/dependency-container": "<1.5",
+    "bit3/contao-dependency-container": "<1.8"
```
Comment (baumannsven) [inline composer.json]: @discordier If replace as self.version, the package not install as symlink. What do you mean?

Comment (discordier) [inline composer.json]: As we are `2.0`, we can scrap up the replacements altogether.

**Comment by discordier** [comment]
I added some more changes.
Please review.

File: src/ContaoServices/ServiceFactory.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+     * @SuppressWarnings(PHPMD.Superglobals)
+     */
+    public function createPageProviderService()
+    {
+        $pageProvider = new PageProvider();
+
+        if (isset($GLOBALS['TL_HOOKS']['getPageLayout']) && is_array($GLOBALS['TL_HOOKS']['getPageLayout'])) {
+            array_unshift(
+                $GLOBALS['TL_HOOKS']['getPageLayout'],
+                [PageProvider::class, 'setPage']
```
Comment (dmolineus) [inline src/ContaoServices/ServiceFactory.php]: We could use the service `cca.legacy_dic.contao_page_provider` instead of the class name.

Comment (discordier) [reply src/ContaoServices/ServiceFactory.php]: Will you change it?

**Pull request merged**